modules/sharedaddy/sharing.js

```diff
@@ -1,364 +1,474 @@
 /* global WPCOM_sharing_counts, grecaptcha */
-var sharing_js_options;
-if ( sharing_js_options && sharing_js_options.counts ) {
-	var WPCOMSharing = {
-		done_urls: [],
-		get_counts: function () {
-			var url, requests, id, service, service_request;
-
-			if ( 'undefined' === typeof WPCOM_sharing_counts ) {
-				return;
-			}
 
-			for ( url in WPCOM_sharing_counts ) {
-				id = WPCOM_sharing_counts[ url ];
+// NOTE: This file intentionally does not make use of polyfills or libraries,
+// including jQuery. Please keep all code as IE11-compatible vanilla ES5, and
+// ensure everything is inside an IIFE to avoid global namespace pollution.
+// Code follows WordPress browser support guidelines. For an up to date list,
+// see https://make.wordpress.org/core/handbook/best-practices/browser-support/
 
-				if ( 'undefined' !== typeof WPCOMSharing.done_urls[ id ] ) {
-					continue;
-				}
+( function () {
+	var currentScript = document.currentScript;
 
-				requests = {
-					// Pinterest handles share counts for both http and https
-					pinterest: [
-						window.location.protocol +
-							'//api.pinterest.com/v1/urls/count.json?callback=WPCOMSharing.update_pinterest_count&url=' +
-							encodeURIComponent( url ),
-					],
-					// Facebook protocol summing has been shown to falsely double counts, so we only request the current URL
-					facebook: [
-						window.location.protocol +
-							'//graph.facebook.com/?callback=WPCOMSharing.update_facebook_count&ids=' +
-							encodeURIComponent( url ),
-					],
-				};
-
-				for ( service in requests ) {
-					if ( ! jQuery( 'a[data-shared=sharing-' + service + '-' + id + ']' ).length ) {
-						continue;
-					}
+	// -------------------------- UTILITY FUNCTIONS -------------------------- //
```

Comment:
As counter-intuitive as this sounds, I would like to create a separate PR (from which this PR is branched) which puts any shareable utility functions into a separate non-jQuery dependency. 

For example, in my carousel PR, I created a (badly-named, messy) library called `jpQuery` which included anything that was more than a 3-liner to replace.

https://github.com/Automattic/jetpack/pull/16420/files#diff-c2ea7f98aaf08faa387719e495ccff36

It seems to be that we are going to have a hard time ripping out jQuery without replacing _some_ non-trivial functions with a shared JS library, though hopefully one that's much smaller than jQuery.

What do you think of that idea? Would you be willing to create a PR that combines your utility functions with jpQuery and then rebase this PR on top of it? I would then do the same with my own PR.

Reply:
@gravityrail I fully agree we need some common base, I'm just not sure what the best mechanism to achieve it is.

These are some of the options I was considering:

1. We can either create a small common library like you suggest and enqueue it async / deferred,
2. or create / assemble (existing micro libraries) a collection of tools / methods and selectively `import` from it,
3. or we can introduce babel, write shiny ES6+ and inject a battery of polyfills for IE10+ under a conditional comment.

Actually, 2 + 3 could be combined. Also either one requires us adding an additional build step, but it could also lead to a clean code that doesn't depend on anything external. But it introduces complexity, potentially mainly around integration with existing build tools / linters / checks / wpcom.

Happy to follow any decision you'll make, just let me know and I can help out.

Reply:
Thank you, @dero and @gravityrail! I fully expected these sorts of concerns to come from an initial PR, where the approach hasn't been defined yet. For the record, I don't mind changing everything altogether or throwing out the code and starting anew; sometimes you have to try something before deciding it's not the right direction 😄 

So with that in mind, and taking other comments into mind here, I think @dero's three options above illustrate the possible paths we can take well.

Starting with option 1, this appears to be the easiest (but not necessarily best) path forward. We create a new library, enqueue it with WP's usual mechanisms, and use it where needed. Everything stays ES5 and IE11-compatible. The risk is that the library becomes a large monolith over time (probably never as big as jQuery, but still), with many different scripts depending on it. The upside is that even if that happens, we're in control of it, so we could e.g. at some point make it smaller by dropping IE11 compatibility, while maintaining the same external API.

Options 2 and 3 involve an entirely different approach, where code would be ES2015+ before being transpiled, and we could take advantage of modules for selective importing from a shared library (or two, or three), not to mention the much better developer experience. This approach breaks down into two further possibilities: either we ship a single ES5 IE11-compatible bundle to everyone, or we do differential serving based on the user's browser. Let's look at each of them separately:

**IE11-compatible bundle for everyone**

This is probably the easiest option of the two. We would enqueue a polyfill set and the transpiled bundle and call it a day. The downside is that the polyfills will almost certainly end up being larger than jQuery, which feels like running from one problem straight into another.

**Differential serving**

This one gets trickier, because as far as I know, we can't really rely on `enqueue` any more.

The general idea is that we'd prepare two bundles, one for old browsers and one for newer browsers, with an arbitrarily defined cutoff point, and we'd serve the right one to each browser.

That could be done with either a server-side mechanism (by looking at the user agent string and guessing which file to serve), or through a client-side mechanism, for example taking advantage of the module/nomodule trick (here's [an example article introducing the approach](https://dev.to/thejohnstew/differential-serving-3dkf)). I would recommend the latter for Jetpack.

Of course, this means we need to inject both a `<script nomodule src="..">` and a `<script type="module" src="...">` into the page, which AFAIK isn't really doable with `enqueue`, though I'm certainly not an expert there 🙂 

Note that even in this approach, we'd need to be vigilant about what we include, since if we're transpiling to this arbitrary point in time (in the case of `module`/`nomodule`, that would be the available feature set circa Chrome 61 / Safari 11), anything else would still need to be polyfilled. Still, it is an excellent path forward for maintaining compatibility on old browsers, improving performance on new ones, and doing all of this without major infrastructure changes to WP.

**Next steps**

So this looks like a key decision that needs to be made before any further work 😄 I'm happy to move this discussion elsewhere or schedule a call if you prefer, but I'm also happy to continue the discussion here. There are a number of possibilities as discussed above, and I'm happy to provide more details on all of them!

Reply:
> [...] battery of polyfills for IE10+ under a conditional comment.

After all these years and I still think conditional comments are a thing. #dinosaurInARoom 🦕 

Should we decide for transpilation, I agree with @sgomes that any bundle of polyfills need to be loaded selectively only on browsers that require them. I'm not a huge fan of the `module` / `nomodule` approach, however, because it's a hack.

I would rather do feature detection ([ES6](https://kangax.github.io/compat-table/es6/), [ES2016+](https://kangax.github.io/compat-table/es2016plus/)). With [Firefox implementing a new, ES2018 compatible RegExp engine in SpiderMonkey](https://www.mozilla.org/en-US/firefox/78.0/releasenotes/#note-788331), all current versions of modern browsers now basically support ES2019 and it shouldn't be too difficult to construct a compound check that only ES2019 (and lower) capable browsers satisfy and only load the polyfill bundle for older browsers. 

If that turns out not being possible for some reason, I wouldn't shy away from user-agent detection on the client. We're basically polyfilling only for Internet Explorer, all other browsers receive frequent updates and their feature set is quite comparable.

Reply:
I'm not a huge fan of the `module` / `nomodule` approach either, trust me 😄 Unfortunately, it's the only "free performance" option we have on the client.

Feature detection is certainly possible, but it adds a roundtrip cost to the scripts, which may or may not be acceptable. To clarify, with feature detection you need to wait for the feature detection script to load and run, determine which file is needed, and then wait for that file to be downloaded and run as well. This is perfectly acceptable for features like what this PR handles (share buttons), but may not be acceptable for any critical JS. Even for non-critical JS, it may make it hard to run in a specific sequence, if that's needed.

I'd shy away from server-side UA string matching if at all possible for this use-case. It requires extra server infrastructure that needs to be regularly updated, and UA strings will likely be getting frozen in the future, at least in Chrome.

Reply:
> Feature detection is certainly possible, but it adds a roundtrip cost to the scripts. [...] but may not be acceptable for any critical JS.

I understand the potential restrictions, @sgomes, and I agree with you that feature detection is not a magic silver bullet that would just work everywhere. But we may want to explore this path in more detail, nevertheless, because we don't _need_ it to work everywhere, we need to work in Jetpack and I'd argue we have a lot of knowledge on and control over the scripts that would be subject to these changes.

Here are a questions that may expand our options:

1. Is there a piece of JS in the Jetpack codebase that would be "critical" enough that it couldn't be delayed? (Any inline JS excluded, those are likely going to remain ES5 compatible.)
2. Are the pieces of JS we're stripping jQuery from interdependent? They usually live in modules that can be enabled independently, so I'm inclined to say "mostly no".
3. Is there any external JS Jetpack runs that needs to run synchronously? We have been working on improving other aspects of Jetpack's performance earlier this year and I don't recall that being the case.

Can we make the ES6 a default that falls back to IE11-compat _on the client_? Maybe something like this could work? It's a bit... let's say _experimental_, but maybe worth a discussion:

```js
// Top of the page. Could be further optimized to init Jetpack_JS async if necessary.
var Jetpack_JS = {
  queue: function( name, fn ) {
    // Runs and memoizes a test checking the current browser ES version capabilities.
    // If the test fails, `loadScript`s an `es5` version of the JS and overrides the fn. Honors queue order during execution.
    // Otherwise run the original ES2019 code immediately.
  };
};

// In (for example) tiled-gallery.js.
Jetpack_JS.queue( 'tiled-gallery', function() {
  // ES2019 code.
} );
```

This approach would simply run the ES2019 code in modern browsers with a (probably) negligible overhead.

IE11, Blink Edge and any other outdated browsers we may wish to support would suffer a performance hit and would end up double-downloading the JS files, but I'd argue that's not much of an issue, because those are desktop browsers and we're not executing the scripts twice, only downloading. And I suppose our end goal is to _support_ IE11 and Blink Edge functionally, but not necessarily optimize for performance in those outdated browsers.

Reply:
Thanks for the really robust discussion. It helps me a lot to read the pros and cons of each approach. 

Given the options described above, here's what I'd recommend:

- Let's use a build system and transpile everything.
- We really only need to detect one user-agent, which is IE11. We will serve "modern" JS to every other browser. Maybe that browser can't run it, but I'm prepared to wear that cost for the sake of simplicity.
- We build our jquery-replacement utility library (let's call it 'common' for now) as an npm package inside the Jetpack monorepo, and import things like this: `import { forEachNode } from '@jetpack/common'`
- As we convert JS to remove jQuery, we _also_ migrate it to this new transpilation system, which does bundle-splitting so that functions that are common across all features end up in a `vendor.min.js` file of some kind - **PLUS** we do this **TWICE**, once each for the two browser-specific builds I will describe next...
- For IE11, we serve a transpiled ES5 + polyfills bundle
- For **everyone** else, we serve a bundle optimized for the last 2 major versions of Chrome, Safari, Firefox and Edge

So, basically, that's it. Since IE11 is the only holdout on nice things like ES6 classes, and since IE11 needs the vast majority of polyfills, this should mean that all the compatibility garbage is reserved for IE11 users and everyone else gets a much more lean build.

Also, in terms of browser detection, we really _only_ need to do that for IE11. If it introduces a performance penalty, I don't really care since IE11 support will go away at some point anyway, at which point we just delete that config from the build system.

What do you think of that?

Reply:
Thank you both, @dero and @gravityrail! Now that I understand Jetpack's needs a bit better I agree that feature detection sounds like a good idea, and I especially like your idea of zero-cost green path with fallback, @dero! That's a clever approach that I'd definitely love to try 👍 

The baseline of "last 2 major versions of Chrome, Safari, Firefox and Edge" that @gravityrail suggested seems perfect. Whenever possible, we should probably try not to use the absolute latest bells and whistles, but we can be reasonably confident that the vast majority of users will be on those releases.

Taking all of the discussion into account, here's how I'd put everything together:
- Perform a test at the beginning of the script, to see if the browser supports basic ES2015 syntax and any newer features that the script makes use of. If not, go to fallback approach. Crucially, this should happen before the script makes any changes to the page, so we don't have to worry about rolling things back or anything.
- The fallback approach is to stop execution immediately and load the fallback script asynchronously. In order for this to work, we need to ensure that all Jetpack scripts work independently of each other, and in any loading order. I think this might already be the case, right?
- Anything that passes the test for "modern browser" but fails execution later is considered a bug, or a side effect of this approach, which we may or may not want to fix. Reducing the amount of users that fall into this category is why I would suggest testing for newer features as well, instead of just basic ES2015 syntax. Of course, we can make these tests part of `@jetpack/common` to make things easier.

As for how to handle the polyfills, we have two options. The difficult, but more performant option, is to have a separate polyfill script that we load and run before the fallback scripts. Here we run into the trickiness of having to manage execution order.

The much easier option, and the one I'd go for, is to include the polyfills with every Jetpack fallback script, which of course means a lot of duplication if there are a bunch of Jetpack options enabled, but significantly simplifies things on our end. And echoing @gravityrail's approach, I think that's a perfectly acceptable tradeoff; we probably shouldn't spend much time thinking about IE11 perf, and as long as things work at a basic level, that's ok. If users want better performance, they'll need to upgrade their browser.

I'll get started by working on a new PR that reimplements this script for modern browsers; it's going to feel great to rewrite the whole thing to be 2x smaller 😄  Once I'm done with that, we can see which bits seem to be reusable, so that we can pull them out into the utility library that @gravityrail suggested, and we can see about adding the feature-detection bits and coming up with that build process! Happy for someone else to look into that in the meantime too, if they're interested 👍

Reply:
Thank you so much! This feels like a good direction. I really appreciate all your efforts and how you're incorporating feedback into the approach.